### PR TITLE
feat: copy failedplacement from work

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
### Description of your changes

Fixes #

I have:

- [ ] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/controllers/workgenerator/controller.go [hunk 1]
```diff
@@ -608,6 +604,40 @@ func getWorkNamePrefixFromSnapshotName(resourceSnapshot *fleetv1beta1.ClusterRes
 	return fmt.Sprintf(fleetv1beta1.WorkNameWithSubindexFmt, crpName, subIndexVal), nil
 }
 
+// setBindingStatus sets the binding status based on the works
+func setBindingStatus(works map[string]*fleetv1beta1.Work, resourceBinding *fleetv1beta1.ClusterResourceBinding) {
+	bindingRef := klog.KRef(resourceBinding.Namespace, resourceBinding.Name)
+	// try to gather the resource binding applied status if we didn't update any associated work spec this time
```
Comment (zhiying-lin) [inline pkg/controllers/workgenerator/controller.go:610]: need to reset the FailedPlacements.

resourceBinding.Status.FailedPlacements = nil

File: pkg/controllers/workgenerator/controller_test.go [hunk 2]
```diff
@@ -476,3 +477,302 @@ func TestBuildAllWorkAvailableCondition(t *testing.T) {
...
+				},
+			},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			binding := &fleetv1beta1.ClusterResourceBinding{}
+			setBindingStatus(tt.works, binding)
+			got := binding.Status.FailedPlacements
```
Comment (zhiying-lin) [inline pkg/controllers/workgenerator/controller_test.go:763]: should we cover the whole status? as setBindingStatus including setting conditions & failedPlacements.

File: pkg/controllers/workgenerator/controller_test.go [hunk 3]
```diff
@@ -476,3 +477,302 @@ func TestBuildAllWorkAvailableCondition(t *testing.T) {
 		})
 	}
 }
+
+func TestSetBindingStatus(t *testing.T) {
+	tests := map[string]struct {
+		works map[string]*fleetv1beta1.Work
+		want  []fleetv1beta1.FailedResourcePlacement
+	}{
```
Comment (zhiying-lin) [inline pkg/controllers/workgenerator/controller_test.go:485]: add a test, two works are available.

File: pkg/controllers/workgenerator/controller_test.go (hunk 2 above)
Comment (ryanzhang-oss) [reply pkg/controllers/workgenerator/controller_test.go:763]: the reason I only focused on the failePlacment is that there are already tests on the buildAllWorkAppliedCondition and buildAllWorkAvailableCondition

File: pkg/utils/controller/controller.go [hunk 4]
```diff
@@ -14,13 +14,16 @@ import (
 	"time"
 
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (zhiying-lin) [inline pkg/utils/controller/controller.go:18]:
```suggestion
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```

File: pkg/utils/controller/controller.go [hunk 5]
```diff
@@ -344,3 +352,83 @@ func FetchAllClusterResourceSnapshots(ctx context.Context, k8Client client.Clien
 	}
 	return resourceSnapshots, nil
 }
+
+func ExtractFailedResourcePlacementsFromWork(work *fleetv1beta1.Work) []fleetv1beta1.FailedResourcePlacement {
```
Comment (zhiying-lin) [inline pkg/utils/controller/controller.go:356]: add a comment? public func needs a comment.

**Review by zhiying-lin — COMMENTED** [review]
two minor style comments

File: pkg/utils/controller/controller.go [hunk 6]
```diff
@@ -353,6 +353,7 @@ func FetchAllClusterResourceSnapshots(ctx context.Context, k8Client client.Clien
 	return resourceSnapshots, nil
 }
 
+// ExtractFailedResourcePlacementsFromWork extracts the failed resource placements from the work. d
```
Comment (zhiying-lin) [inline pkg/utils/controller/controller.go:356]:
```suggestion
// ExtractFailedResourcePlacementsFromWork extracts the failed resource placements from the work.
```

**Review by zhiying-lin — APPROVED** [review]

**Pull request merged**